validator_worker/src/follower.rs

```diff
@@ -333,3 +333,527 @@ async fn on_error<'a, C: Unlocked + 'static>(
+                }
+            ), "InvalidNewState::RootHash is the rejection reason when the state roots don't match");
+        }
+
+        // - Case where sentry.adapter.verify(…) will return false -> signature is from a different validator
+        {
+            let proposed_balances = get_initial_balances();
+            let state_root = proposed_balances
+                .encode(
+                    channel_context.context.id(),
+                    channel_context.token.precision.get(),
+                )
+                .expect("should encode");
+            let new_state: NewState<UncheckedState> = NewState {
+                state_root: state_root,
+                signature: IDS[&*FOLLOWER].to_checksum(),
+                balances: proposed_balances.into_unchecked(),
+            };
+            let res = on_new_state(
+                &sentry,
+                &channel_context,
+                get_initial_balances(),
+                new_state,
+                UnifiedNum::from_u64(4000),
+            )
+            .await
+            .expect("Shouldn't return an error");
+            assert!(matches!(
+                res,
+                ApproveStateResult::RejectedState {
+                    reason: InvalidNewState::Signature,
+                    ..
+                }
+            ), "InvalidNewState::Signature is the rejection reason when the signature can't be verified");
+        }
+
+        // - Case where LastApprovedResponse new state balances have a payout mismatch resulting in an error
+        {
+            let mut last_approved_balances = get_initial_balances().into_unchecked();
+            last_approved_balances
+                .earners
+                .insert(*GUARDIAN_2, UnifiedNum::from_u64(10_000));
+            let _mock_guard = server_setup
+                .setup_last_approved_response(last_approved_balances, String::new())
+                .await;
+
+            let proposed_balances = get_initial_balances();
+            let state_root = proposed_balances
+                .encode(
+                    channel_context.context.id(),
+                    channel_context.token.precision.get(),
+                )
+                .expect("should encode");
+            let new_state: NewState<UncheckedState> = NewState {
+                state_root: state_root,
+                signature: IDS[&*LEADER].to_checksum(),
+                balances: proposed_balances.into_unchecked(),
+            };
+            let res = on_new_state(
+                &sentry,
+                &channel_context,
+                get_initial_balances(),
+                new_state,
+                UnifiedNum::from_u64(4000),
+            )
+            .await
+            .expect("Shouldn't return an error");
+            assert!(matches!(
+                res,
+                ApproveStateResult::RejectedState {
+                    reason: InvalidNewState::Transition,
+                    ..
+                }
```

Comment:
Deriving `Eq, PartialEq` won't be very straight-forward for `ApproveStateResult` because of the Error types and I think checking just the reason would be sufficient. We can leave it with `matches` for now.

Reply:
yes, you're right, Sadly reqwest::Error does not implement either of these traits so it will be difficult to do so.
Let's indeed leave it with a `matches!`.